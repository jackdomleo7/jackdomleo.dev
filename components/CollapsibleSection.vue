<template>
  <div :id="`collapsible-section-${id}`" class="collapsible-section">
    <tag :is="headingTag" class="collapsible-section__header">
      <button
        :id="`collapsible-section-${id}__header`"
        :aria-expanded="isOpen.toString()"
        :aria-controls="`collapsible-section-${id}__panel`"
        @click="isOpen = !isOpen"
      >
        <span>{{ heading }}</span>
        <svg-icon name="chevron-down" />
      </button>
    </tag>
    <section
      :id="`collapsible-section-${id}__panel`"
      class="collapsible-section__body"
      :class="{'collapsible-section__body--hidden': !isOpen}"
      :aria-labelledby="`collapsible-section-${id}__header`"
    >
      <div class="collapsible-section__body-inner">
        <slot />
      </div>
    </section>
  </div>
</template>

<script lang="ts">
import Vue, { PropType } from 'vue'
import { stringHasNoWhitespace } from '@/helpers/propValidators'

export default Vue.extend({
  name: 'CollapsibleSection',
  props: {
    headingTag: {
      type: String as PropType<'h2'|'h3'|'h4'|'h5'|'h6'>,
      required: true,
      validator: (value: string): boolean => {
        return ['h2', 'h3', 'h4', 'h5', 'h6'].includes(value)
      }
    },
    id: {
      type: String,
      required: true,
      validator: (value: string): boolean => {
        return stringHasNoWhitespace(value)
      }
    },
    heading: {
      type: String,
      required: true
    }
  },
  data () {
    return {
      isOpen: false
    }
  }
})
</script>

<style lang="scss" scoped>
.collapsible-section {
  padding: 0;
  width: 100%;
  background-color: var(--colour-foreground-1);
  border-radius: 0.75rem;
  box-shadow: var(--shadow-standard);

  &__header {
    margin: 0;

    button {
      display: flex;
      align-items: center;
      justify-content: space-between;
      color: var(--colour-text-primary);
      padding: 0.875rem 1rem;
      border: 0;
      background-color: transparent;
      width: 100%;
      cursor: pointer;
      font-size: var(--text-heading);
      text-align: left;

      @media (min-width: $responsive-standard-tablet) {
        padding: 1.5rem;
      }

      &:focus {
        outline: none;

        svg {
          outline: 1px solid;
          outline-color: Highlight;
          outline-color: -webkit-focus-ring-color;
          outline-offset: 1px;
        }
      }

      &[aria-expanded="true"] {
        svg {
          transform: rotate(180deg);
        }
      }
    }

    span {
      flex: 1;
      padding-right: 1rem;
    }

    svg {
      height: 3rem;
      width: 3rem;
      transition: transform 260ms ease;
      will-change: transform;
    }
  }

  &__body {
    font-size: var(--text-body);
    transition: max-height 500ms ease;
    will-change: max-height;
    overflow: hidden;
    max-height: 50rem;

    &-inner {
      padding: 0.125rem 1rem 1.25rem 1rem;

      @media (min-width: $responsive-standard-tablet) {
        padding: 0.125rem 1.5rem 1.5rem 1.5rem;
      }
    }

    &--hidden {
      max-height: 0;
    }

    ::v-deep p {
      margin-block: 0.5rem;

      &:first-of-type {
        margin-top: 0;
      }

      &:last-of-type {
        margin-bottom: 0;
      }
    }
  }
}
</style>
