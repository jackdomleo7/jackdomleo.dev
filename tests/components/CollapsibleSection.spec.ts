import { mount, shallowMount } from '@vue/test-utils'
import { configureAxe, toHaveNoViolations } from 'jest-axe'
import CollapsibleSection from '@/components/CollapsibleSection.vue'

expect.extend(toHaveNoViolations)

const axe = configureAxe({
  rules: {
    region: { enabled: false } // Don't need to check because components are loaded in isolation
  }
})

describe('CollapsibleSection', () => {
  /**
   * Vue instance
   */

  it('is a Vue instance', () => {
    const wrapper = mount(CollapsibleSection, {
      propsData: {
        heading: 'What is the meaning of life?',
        headingTag: 'h2',
        id: 'faq-1'
      },
      slots: {
        default: '<p>The meaning of life is 42.</p>'
      },
      stubs: ['svg-icon']
    })

    expect(wrapper.vm).toBeTruthy()
  })

  /**
   * Snapshots
   */

  it('matches snapshot when mounted', () => {
    const wrapper = mount(CollapsibleSection, {
      propsData: {
        heading: 'What is the meaning of life?',
        headingTag: 'h2',
        id: 'faq-1'
      },
      slots: {
        default: '<p>The meaning of life is 42.</p>'
      },
      stubs: ['svg-icon']
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  it('matches snapshot when shallow mounted', () => {
    const wrapper = shallowMount(CollapsibleSection, {
      propsData: {
        heading: 'What is the meaning of life?',
        headingTag: 'h2',
        id: 'faq-1'
      },
      slots: {
        default: '<p>The meaning of life is 42.</p>'
      },
      stubs: ['svg-icon']
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  /**
   * A11y
   */

  it('to have no obvious accessibility violations', async () => {
    const wrapper = mount(CollapsibleSection, {
      propsData: {
        heading: 'What is the meaning of life?',
        headingTag: 'h2',
        id: 'faq-1'
      },
      slots: {
        default: '<p>The meaning of life is 42.</p>'
      },
      stubs: ['svg-icon']
    })

    await expect(axe(wrapper.element)).resolves.toHaveNoViolations()
  })

  /**
   * Props
   */

  const headingTags = ['h2', 'h3', 'h4', 'h5', 'h6']
  for (const headingTag of headingTags) {
    it(`sets the <${headingTag}></${headingTag}> heading tag when set`, () => {
      const wrapper = shallowMount(CollapsibleSection, {
        propsData: {
          heading: 'What is the meaning of life?',
          headingTag,
          id: 'faq-1'
        },
        slots: {
          default: '<p>The meaning of life is 42.</p>'
        },
        stubs: ['svg-icon']
      })

      expect(wrapper.find(`${headingTag}.collapsible-section__header`).exists()).toBe(true)
    })
  }
  
  it('sets the heading of the collapsible section when set', () => {
    const wrapper = shallowMount(CollapsibleSection, {
      propsData: {
        heading: 'What is the meaning of life?',
        headingTag: 'h2',
        id: 'faq-1'
      },
      slots: {
        default: '<p>The meaning of life is 42.</p>'
      },
      stubs: ['svg-icon']
    })

    expect(wrapper.find('h2.collapsible-section__header').text()).toBe('What is the meaning of life?')
  })

  it('sets the id of the collapsible section', () => {
    const wrapper = shallowMount(CollapsibleSection, {
      propsData: {
        heading: 'What is the meaning of life?',
        headingTag: 'h2',
        id: 'faq-1'
      },
      slots: {
        default: '<p>The meaning of life is 42.</p>'
      },
      stubs: ['svg-icon']
    })

    expect(wrapper.find('div#collapsible-section-faq-1').exists()).toBe(true)
    expect(wrapper.find('button#collapsible-section-faq-1__header').exists()).toBe(true)
    expect(wrapper.find('button[aria-controls="collapsible-section-faq-1__panel"]').exists()).toBe(true)
    expect(wrapper.find('section#collapsible-section-faq-1__panel').exists()).toBe(true)
    expect(wrapper.find('section[aria-labelledby="collapsible-section-faq-1__header"]').exists()).toBe(true)
  })

  /**
   * Slots
   */

  it('sets the default slot', () => {
    const wrapper = shallowMount(CollapsibleSection, {
      propsData: {
        heading: 'What is the meaning of life?',
        headingTag: 'h2',
        id: 'faq-1'
      },
      slots: {
        default: '<p>The meaning of life is 42.</p>'
      },
      stubs: ['svg-icon']
    })

    expect(wrapper.find('div.collapsible-section__body-inner').text()).toBe('The meaning of life is 42.')
  })

  /**
   * Interactivity
   */

  it('can be opened and closed', async () => {
    const wrapper = shallowMount(CollapsibleSection, {
      propsData: {
        heading: 'What is the meaning of life?',
        headingTag: 'h2',
        id: 'faq-1'
      },
      slots: {
        default: '<p>The meaning of life is 42.</p>'
      },
      stubs: ['svg-icon']
    })

    expect(wrapper.find('button#collapsible-section-faq-1__header').attributes('aria-expanded')).toBe('false')
    expect(wrapper.find('section#collapsible-section-faq-1__panel').attributes('class')).toContain('collapsible-section__body--hidden')

    await wrapper.find('button#collapsible-section-faq-1__header').trigger('click')

    expect(wrapper.find('button#collapsible-section-faq-1__header').attributes('aria-expanded')).toBe('true')
    expect(wrapper.find('section#collapsible-section-faq-1__panel').attributes('class')).not.toContain('collapsible-section__body--hidden')

    await wrapper.find('button#collapsible-section-faq-1__header').trigger('click')

    expect(wrapper.find('button#collapsible-section-faq-1__header').attributes('aria-expanded')).toBe('false')
    expect(wrapper.find('section#collapsible-section-faq-1__panel').attributes('class')).toContain('collapsible-section__body--hidden')
  })
})
