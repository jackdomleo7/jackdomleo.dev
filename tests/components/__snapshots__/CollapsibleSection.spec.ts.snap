// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollapsibleSection matches snapshot when mounted 1`] = `
<div
  class="collapsible-section"
  id="collapsible-section-faq-1"
>
  <h2
    class="collapsible-section__header"
  >
    <button
      aria-controls="collapsible-section-faq-1__panel"
      aria-expanded="false"
      id="collapsible-section-faq-1__header"
    >
      <span>
        What is the meaning of life?
      </span>
       
      <svg-icon-stub
        name="chevron-down"
      />
    </button>
  </h2>
   
  <section
    aria-labelledby="collapsible-section-faq-1__header"
    class="collapsible-section__body collapsible-section__body--hidden"
    id="collapsible-section-faq-1__panel"
  >
    <div
      class="collapsible-section__body-inner"
    >
      <p>
        The meaning of life is 42.
      </p>
    </div>
  </section>
</div>
`;

exports[`CollapsibleSection matches snapshot when shallow mounted 1`] = `
<div
  class="collapsible-section"
  id="collapsible-section-faq-1"
>
  <h2
    class="collapsible-section__header"
  >
    <button
      aria-controls="collapsible-section-faq-1__panel"
      aria-expanded="false"
      id="collapsible-section-faq-1__header"
    >
      <span>
        What is the meaning of life?
      </span>
       
      <svg-icon-stub
        name="chevron-down"
      />
    </button>
  </h2>
   
  <section
    aria-labelledby="collapsible-section-faq-1__header"
    class="collapsible-section__body collapsible-section__body--hidden"
    id="collapsible-section-faq-1__panel"
  >
    <div
      class="collapsible-section__body-inner"
    >
      <p>
        The meaning of life is 42.
      </p>
    </div>
  </section>
</div>
`;
